### PR TITLE
Handle EventStream canary events gracefully

### PR DESCRIPTION
**Pull request by MobyDickIsAfrican**
Ticket: https://phabricator.wikimedia.org/T349697

* Added hasCanary method to schema interface
* Implemented hasCanary on each event type

### CODE REVIEW

File: schema.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ import (
 type schema interface {
 	unmarshal(evt *Event) error
 	timestamp() time.Time
+	hasCanaryEvent() bool
```
Comment (ruairi-wmf) [inline schema.go]: Should we split this interface into individual interfaces and then combine it with composition?  Like we do with other interfaces?

File: client.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ func (cl *Client) PageCreate(ctx context.Context, since time.Time, handler func(
 			evt := new(PageCreate)
 			parseSchema(evt, msg, store)
 
+			if hasCanaryEvent(evt) {
```
Comment (prabhat393) [inline client.go]: Pls rename all the methods to `isCanaryEvent()`

**Comment by protsack-stephan** [comment]
Looks good, but I wonder can we move this code closer to the root of the package?

The long term problem with this solution that I see is that we have to add it every time we add new stream, and that's error prone, cuz we can always forget.

Can we try to do this filtering somewhere in [here](https://github.com/wikimedia-enterprise/wmf-event-stream-sdk-go/blob/main/subscribe.go#L51)?

**Comment by WikiCreatorOfWorlds** [comment]
Makes sense. I've moved the logic to the subscribe function.

File: subscribe.go [hunk 3]
```diff
@@ -47,6 +48,12 @@ func subscribe(ctx context.Context, client *http.Client, url string, since time.
 		}
 
 		if len(evt.ID) > 0 && len(evt.Data) > 0 && err == nil {
+			bsd := new(baseData)
+			json.Unmarshal(evt.Data, bsd)
+
+			if bsd.Meta.Domain == "canary" {
+				continue
+			}
 			handler(evt)
```
Comment (protsack-stephan) [inline subscribe.go:60]: I suggest adding an empty line to visually separate the if statement from handler call.

File: subscribe.go [hunk 4]
```diff
@@ -47,6 +48,12 @@ func subscribe(ctx context.Context, client *http.Client, url string, since time.
 		}
 
 		if len(evt.ID) > 0 && len(evt.Data) > 0 && err == nil {
+			bsd := new(baseData)
+			json.Unmarshal(evt.Data, bsd)
```
Comment (protsack-stephan) [inline subscribe.go]: I think we should add a error handling here, can we do something like:

```go
bsd := new(baseData)

if err := json.Unrashal(evt.Data, bsd); err != nil {
  return err
}
```

**Review by protsack-stephan — COMMENTED** [review]
Good work! Couple small comments.

File: subscribe.go (hunk 4 above)
Comment (WikiCreatorOfWorlds) [reply subscribe.go]: Done.

**Comment by prabhat393** [comment]
Looks good. Feel free to merge this.

**Pull request merged**